Surface reparameterization conflict fix

jiyess opened the pull request:
This PR fixes the previous surface reparameterization implementation, resolving a conflict with the current stable version. The updated implementation can be activated by enabling `gsHLBFGS`.

Additionally, a new example named `surfaceReparameterization_example.cpp` has been added.

-----
Please consider the following checklist before issuing a pull
request:
- [X] Have you added an explanation of what your changes do and why
  you'd like us to include them?
- [X] Have you documented any new codes using Doxygen comments?
- [X] Have you written new tests or examples for your changes?
-----

jiyess committed:
:ambulance: fix conflict with stable branch

jiyess committed:
:ambulance: fix conflict with stable branch - push

jiyess committed:
:ambulance: fix conflict with stable branch - push 2

jiyess committed:
:ambulance: fix conflict with stable branch - push test 3

hverhelst commented:
This PR seems to depend on [746](https://github.com/gismo/gismo/pull/746). It would be better to merge that one first since it contains unit tests, examples and XML support.

hverhelst commented on examples/surfaceReparametrization_example.cpp:
```diff
@@ -0,0 +1,48 @@
+/** @file bSplineCurve_example.cpp
```
Change this line

hverhelst commented on src/gsCore/gsBasis.h:
```diff
@@ -719,14 +719,14 @@ class gsBasis : public gsFunctionSet<T>
     /// this basis, that points to the first element of the domain
-    GISMO_DEPRECATED // @hverhelst: this function will be deprecated, since it will now always point to the first element of the domain, hence call this->domain()->beginAll()
-    virtual domainIter makeDomainIterator() const;
+//    GISMO_DEPRECATED // @hverhelst: this function will be deprecated, since it will now always point to the first element of the domain, hence call this->domain()->beginAll()
```
Why is this one (and the next one) commented?

hverhelst commented on src/gsCore/gsBasis.hpp:
```diff
@@ -538,15 +538,15 @@ template<class T>
-typename gsBasis<T>::domainIter
-gsBasis<T>::makeDomainIterator(const boxSide &s) const
-{ return this->domain()->beginBdr(s); }
+//template<class T>
```
Idem

hverhelst commented on src/gsCore/gsComposedBasis.h:
```diff
@@ -0,0 +1,307 @@
+    gsMatrix<T> support() const override
+    {
+        gsMatrix<T> supp = m_basis->support();
+        // gsGridIterator<T,CUBE> pt(supp,math::pow(2,this->domainDim()));
```
We might need to decide on this code: is it needed? Then uncomment. Otherwise, remove

hverhelst commented on src/gsCore/gsComposedBasis.h:
```diff
@@ -0,0 +1,307 @@
+        result = tmp;
+    }
+
+    // void control_deriv_into(const gsMatrix<T> & points, gsMatrix<T> & result)
```
Remove if not needed

jiyess replied:
Already merged via PR #746  — removing this file from the branch.

hverhelst commented on src/gsCore/gsComposedGeometry.h:
```diff
@@ -0,0 +1,177 @@
+
+    Author(s):
+        H.M. Verhelst   (2019-..., TU Delft)
+        A. Mantzaflaris (2019-..., Inria)
```
Change

hverhelst commented on src/gsCore/gsComposedGeometry.h:
```diff
@@ -0,0 +1,177 @@
+/** @file gsComposedGeometry.h
+
+    @brief
```
Add

hverhelst commented on src/gsCore/gsComposedBasis.h:
```diff
@@ -0,0 +1,307 @@
+/** @file gsComposedBasis.h
+
+    @brief
```
Add

hverhelst commented on src/gsCore/gsComposedBasis.h:
```diff
@@ -0,0 +1,307 @@
+
+    Author(s):
+        H.M. Verhelst   (2019-..., TU Delft)
+        A. Mantzaflaris (2019-..., Inria)
```
Change

hverhelst commented on src/gsCore/gsComposedGeometry.h:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    /**
+     * @brief      XXXX
```
Add

hverhelst commented on src/gsCore/gsComposedGeometry.h:
```diff
@@ -0,0 +1,177 @@
+    { }
+
+    /**
+     * @brief      XXXX
```
Add

hverhelst commented on src/gsMSplines/gsMappedBasis.h:
```diff
@@ -210,9 +210,9 @@ class gsMappedBasis  : public gsFunctionSet<T>
-    GISMO_DEPRECATED
-    typename gsBasis<T>::domainIter makeDomainIterator(const index_t patch,const boxSide & s) const
-    { return m_bases[patch]->makeDomainIterator(s); }
+//    GISMO_DEPRECATED
```
Why commented?

hverhelst commented on src/gsMSplines/gsMappedSingleBasis.h:
```diff
@@ -327,19 +327,19 @@ class gsMappedSingleBasis : public gsBasis<T>
-        // TODO Not always working: make it more general
-        return m_basis->getBase(m_index).makeDomainIterator(s);
-    }
+//    GISMO_DEPRECATED
```
Why commented?

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+
+	gsOptionList m_options;
+
+	T m_lambda1 = 1.0, m_lambda2 = 1.0;
```
Initialize values in the constructor

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+  template<class T = real_t>
+  gsMultiPatch<T> convertIntoBSpline(const gsMultiPatch<T>& mp, const gsMatrix<T>& coefsMobiusIn) {
+	gsMultiPatch<T> result;
+
```
Add an assert for the dimensions, since the rest is hard coded for domain dimension 2

jiyess replied:
Done! Thanks.

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+	// Precompute Mobius domain matrix (use structured bindings in C++17)
+	gsAsConstVector<T> coefsMobius(coefsMobiusIn.data(), 4);
+	gsMatrix<T, 2, 2> alpha;
+	alpha << coefsMobius(0), coefsMobius(2), coefsMobius(1), coefsMobius(3);
```
Can this be done with some alpha=coefsMobius.reshape(..)?

jiyess replied:
Yes. Thanks!

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+	  patch->eval_into(xieta, eval_geo);
+
+	  // Convert the patch's basis into a tensor B-spline basis
+	  auto& bbasis = static_cast<gsTensorBSplineBasis<2, T>&>(patch->basis());
```
Remove auto (as the type is very clear)

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  template<class T = real_t>
+  gsMultiPatch<T> convertIntoBSpline(const gsMultiPatch<T>& mp, const gsMatrix<T>& coefsMobiusIn) {
```
The name of this function suggests something exact. However, it's approximate. Maybe the name can be chosen differently.
In addition, why using fitting instead of L2 projection, quasi interpolation etc? Is there a clear preference?

jiyess replied:
The function convertIntoBSpline is now renamed by approximateWithBSpline.

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+		: m_mp(patches) {
+	  gsMatrix<T, 2, 2> alpha;
+	  alpha.setConstant(0.5);
+	  m_mobiusDomain = gsMobiusDomain<2, T>(alpha);
```
Maybe gsMobiusDomain<2,T>(gsMatrix<T,2,1>::Constant(2)) works?

jiyess replied:
Indeed, `m_mobiusDomain = gsMobiusDomain<2,T>(gsMatrix<T,2,2>::Constant(0.5));` works. Thanks!

hverhelst commented on src/gsModeling/gsSurfaceReparameterization.h:
```diff
@@ -0,0 +1,205 @@
+	  initialGuessVector.setConstant(0.5); // Adjust this as necessary for better performance
+
+	  // Set up the optimizer
+	  gsHLBFGS<T> optimizer(&objFuncSurface);
```
Can we make the optimizer a member of the class (a pointer to gsOptimizer) or a template? In that way, we can provide the freedom to the user to select the optimizer and we don't have to disable the class without HLBFGS. 
The optimizer can be given as an argument to the constructor, and by default giving gradient descent.

hverhelst replied:
We did it here:
https://github.com/gismo/gismo/blob/b275730cf650745ebca6c845ebf3c1ff6b0b0f61/src/gsAssembler/gsAdaptiveParametrization.h#L139

hverhelst commented on src/gsNurbs/gsMobiusDomain.h:
```diff
@@ -0,0 +1,183 @@
+/** @file gsMultiBasis.h
```
Change

hverhelst commented on src/gsNurbs/gsMobiusDomain.h:
```diff
@@ -0,0 +1,183 @@
+{
+
+template <short_t DIM, class T>
+class gsMobiusDomain : public gsFunction<T>
```
Remove commented functions

hverhelst commented on src/gsNurbs/gsMobiusDomain.h:
```diff
@@ -0,0 +1,183 @@
+    m_alpha = alpha;
+  }
+
+  void updateGeom(const gsAsConstVector<T>& alpha) {
```
This seems redundant with the previous one

hverhelst commented on src/gsNurbs/gsSquareDomain.h:
```diff
@@ -0,0 +1,216 @@
+/** @file gsSquareDomain.h
```
This file does not seem to be needed in this PR, correct? 
There is another one upcoming in another branch (Adaptive parametrization)

jiyess replied:
Removed. Thanks!

hverhelst reviewed (commented):
Left some comments, please merge this one after [746](https://github.com/gismo/gismo/pull/746)

filiatra commented on src/gismo.h:
```diff
@@ -81,6 +81,9 @@ namespace internal
 
+#include <gsCore/gsComposedBasis.h>
+#include <gsCore/gsComposedGeometry.h>
+
```
these files are not yet in stable.
Is the plan that they are added in this PR  ?

hverhelst committed:
Merge remote-tracking branch 'origin/stable' into surface-reparameterization-conflict3

hverhelst committed:
cleanup

hverhelst commented on src/gsNurbs/gsMobiusDomain.h:
We might need to change the name, since `gs...Domain` is reserved for everything in the `gsDomain/` directory
@filiatra  what do you think?

filiatra replied:
Yes, we could call it gsMobiusMap

hverhelst reviewed (commented):
I did some cleaning, but there are small comments left @jiyess

jiyess committed:
:art: add dimension checks for 2D domain and 3D geometry compatibility for the convertIntoBSpline function

jiyess committed:
:art: Replace manual assignment with reshape for Mobius alpha matrix in the convertIntoBSpline function

jiyess committed:
:art: Rename convertIntoBspline with approximateWithBSpline

jiyess committed:
:art: Replace gsMatrix<T, 2, 2> alpha; alpha.setConstant(0.5); m_mobiusDomain = gsMobiusDomain<2, T>(alpha); with m_mobiusDomain = gsMobiusDomain<2,T>(gsMatrix<T,2,2>::Constant(0.5));

jiyess committed:
:fire: remove gsSquareDomain.h from this PR

hverhelst committed:
Rename `gsMobiusDomain` -> `gsMobiusMap`
Add default constructor for `gsOptimizer` classes
Add `setProblem` to `gsOptimizer` classes
Add `gsOptimizer` as argument to `gsSurfaceReparametrization` to avoid HLBFGS (which is optional) by default

filiatra reviewed (changes requested):
Please merge latest stable so that the tests will run, and then we can merge it !

hverhelst committed:
Merge remote-tracking branch 'origin/stable' into surface-reparameterization-conflict3

hverhelst committed:
add missing file

hverhelst commented:
@jiyess can you check the example with HLBFGS disabled? Then it takes gradient descent by default. I hope it works

filiatra commented on src/gsNurbs/gsMobiusMap.h:
```diff
@@ -0,0 +1,131 @@
+ protected:
+//  T m_alpha1, m_alpha2, m_beta1, m_beta2;
+  gsMatrix<T> m_alpha;
+  gsTensorBSpline<DIM,T> m_domain;
```
Change to gsBasis<T> *

filiatra commented on src/gsNurbs/gsMobiusMap.h:
```diff
@@ -0,0 +1,131 @@
+namespace gismo
+{
+
+template <short_t DIM, class T>
```
remove DIM, seems to not be needed

hverhelst committed:
remove DIM and m_domain from MobiusMap

jiyess commented:
> @jiyess can you check the example with HLBFGS disabled? Then it takes gradient descent by default. I hope it works

@hverhelst Yes, it works on my macOS machine with HLBFGS disabled. Thanks a lot for your help!

hverhelst commented:
Thanks for checking! Then we merge now
